File: frontend/src/app/candidate.service.ts
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { Candidate } from './Model/candidate';
import { environment } from 'src/environments/environment';
import { Skill } from './Model/Skill';
@Injectable({
  providedIn: 'root',
})
export class CandidateService {
  private apiServerUrl = environment.apiBaseUrl;
  constructor(private http: HttpClient) {}

  public getCandidates(): Observable<Candidate[]> {
    return this.http.get<any>(`${this.apiServerUrl}/api/candidates`);
  }
  public addCandidate(candidate: Candidate): Observable<Candidate> {
    return this.http.post<Candidate>(
      `${this.apiServerUrl}/api/candidates`,
      candidate
    );
  }

  public removeSkillFromCandidate(
    idCandidate: number,
    idSkill: number
  ): Observable<Candidate> {
    return this.http.delete<any>(
      `${this.apiServerUrl}/api/candidates/removeSkill/${idCandidate}/${idSkill}`
    );
  }

  public removeCandidate(id: number): Observable<Candidate> {
    return this.http.delete<any>(`${this.apiServerUrl}/api/candidates/${id}`);
  }

  public searchCandidatesByName(name: {}): Observable<Candidate[]> {
    return this.http.get<any>(
      `${this.apiServerUrl}/api/candidates/findByName`,
      { params: name }
    );
  }

  public searchCandidateBySkills(parameters: []): Observable<Candidate[]> {
    let params = new HttpParams();
    params = params.append('numOfParams', parameters.length.toString());
    parameters.forEach((skill: Skill, index: number) => {
      params = params.append(
        'Skill' + (index + 1).toString(),
        skill.id.toString()
      );
      console.log(params);
    });

    return this.http.get<any>(`${this.apiServerUrl}/api/candidates`, {
      params: params,
    });
  }

  public addSkillToCandidate(
    idCandidate: number,
    skill: Skill
  ): Observable<any> {
    return this.http.put<Candidate>(
      `${this.apiServerUrl}/api/candidates/addSkill/${idCandidate}`,
      skill
    );
  }
}

File: src/main/java/com/example/codingTaskIntership/service/CandidateService.java
package com.example.codingTaskIntership.service;

import java.util.List;
import java.util.Optional;

import com.example.codingTaskIntership.model.Candidate;
import com.example.codingTaskIntership.model.Skill;

public interface CandidateService {
	
	
	public List<Candidate> findAll();
	public Optional<Candidate> findOne(Long id);
	public Candidate addCandidate (Candidate candidate);
	public Optional<Candidate> delete(Long id);
	public Candidate addSkill (Long idCandidate, Long idSkill);
	public Candidate removeSkill (Long idCandidate, Long idSkill);
	public Candidate save (Candidate candidate);
	public List<Candidate> findByName(String name);
	public List<Candidate> findBySkill(Long idSkill1, Long idSkill2, Long idSkill3, Long numOfParams);

}
